### PR TITLE
Dataset inspection functionality

### PR DESCRIPTION
**Pull request by billschereriii**
Add dataset inspection capabilities, including:

- Get a list of metadata field names
- Get the type of a metadata field
- Get a list of tensor names
- Get the type of a tensor

Support added for C++, Python, C, and Fortran, though Fortran support the two list functions is deferred to a future PR due to the inherent difficulty in Fortran of working with arrays of C style strings

### CODE REVIEW

**Review by MattToast — COMMENTED** [review]
Did an in depth look at the python/pybind/Py*.cpp and everything looks great!! Peaked through everything else (aside of Fortran) and didn't see anything that would immediately raise any red flags.

Overall LGTM!! Withholding approval simply to allow others time to review.

File: src/c/c_dataset.cpp [hunk 1]
```diff
@@ -337,3 +337,113 @@ SRError get_meta_strings(void* dataset,
...
+    SRSetLastError(e);
+    result = e.to_error_code();
+  }
+  catch (...) {
+    SRSetLastError(SRInternalException("Unknown exception occurred"));
+    result = SRInternalError;
+  }
+
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/c/c_dataset.cpp:449]: It's preferred, but not mandatory. I'll add one

[the remark this answers is not in the document]
Comment (billschereriii) [reply src/c/c_dataset.cpp]: Added

**Review by mellis13 — CHANGES_REQUESTED** [review]
Great addition to the library!  Just a couple of small comments/questions.

File: include/dataset.h [hunk 2]
```diff
@@ -275,6 +275,18 @@ class DataSet : public SRObject
...
+        *   \brief Retrieve the name of the DataSet
+        *   \returns The name of the DataSet
+        */
+        std::string get_name() const { return _dsname; }
+
+        /*!
+        *   \brief Change the name for the DataSet
+        *   \param name The name for the DataSet
+        */
+        void set_name(std::string name) { _dsname = name; }
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/dataset.h:288]: Not sure if it matters but we certainly could

File: include/dataset.h [hunk 3]
```diff
@@ -283,16 +295,51 @@ class DataSet : public SRObject
...
+        void get_metadata_field_names(char**& data,
+                                      size_t& n_strings,
+                                      size_t*& lengths);
+
+        /*!
+        *   \brief Retrieve the data type of a metadata field in the DataSet
+        *   \param name The name of the metadata field
+        *   \returns The data type for the metadata field
+        */
+        SRMetaDataType get_metadata_field_type(const std::string& name);
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/dataset.h:342]: I believe they do. I'll add throws documentation

File: src/cpp/dataset.cpp [hunk 4]
```diff
@@ -218,9 +218,27 @@ std::vector<std::string> DataSet::get_tensor_names()
...
+{
+    // Track calls to this API function
+    LOG_API_FUNCTION();
+
+    if (_metadata.has_field(".tensor_names")) {
+        _metadata.get_string_values(
+            ".tensor_names", data, n_strings, lengths);
+    }
+    else {
+        *data = NULL;
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/dataset.cpp:233]: Probably, will do

File: src/cpp/dataset.cpp [hunk 5]
```diff
@@ -218,9 +218,27 @@ std::vector<std::string> DataSet::get_tensor_names()
...
 }
 
+// Retrieve tensor names from the DataSet
+void DataSet::get_tensor_names(
+    char**& data, size_t& n_strings, size_t*& lengths)
+{
+    // Track calls to this API function
+    LOG_API_FUNCTION();
+
+    if (_metadata.has_field(".tensor_names")) {
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/dataset.cpp:228]: The existence of the `.tensor_names` field implied to me that there was a performance or other concern with interrogating the tensorpack, which is why I used that source. Open to discussion on this point if you feel strongly that the implementation should be changed

File: include/c_dataset.h [hunk 6]
```diff
@@ -224,6 +224,52 @@ SRError get_meta_strings(void* dataset,
...
+*/
+SRError get_metadata_field_names(
+    void* dataset, char*** data, size_t* n_strings, size_t** lengths);
+
+/*!
+*   \brief Retrieve the data type of a metadata field in the DataSet
+*   \param dataset The dataset to use for this operation
+*   \param name The name of the metadata field (null-terminated string)
+*   \param name_len The length in bytes of the metadata field name
+*   \param ttype Receives the type for the specified metadata field
```
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/c_dataset.h:267]: fixed

File: include/dataset.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/dataset.h:288]: Added an exception

File: include/dataset.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply include/dataset.h:342]: Actually, get_tensor_type() would have just crashed.

File: src/cpp/dataset.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (billschereriii) [reply src/cpp/dataset.cpp:228]: Nothing actionable here short of eliminating the duplication, but that's high risk and I'll move that to a different issue for separate resolution

**Review by ashao — APPROVED** [review]
Fortran implementation is good to go! No comments from me

**Review by mellis13 — APPROVED** [review]
LGTM -- thanks!

**Pull request merged**